besu/src/main/java/org/hyperledger/besu/cli/subcommands/storage/TrieLogHelper.java

```diff
@@ -0,0 +1,359 @@
+/*
+ * Copyright contributors to Hyperledger Besu.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.hyperledger.besu.cli.subcommands.storage;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static org.hyperledger.besu.controller.BesuController.DATABASE_PATH;
+
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.ethereum.chain.Blockchain;
+import org.hyperledger.besu.ethereum.chain.MutableBlockchain;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.trie.bonsai.storage.BonsaiWorldStateKeyValueStorage;
+import org.hyperledger.besu.ethereum.worldstate.DataStorageConfiguration;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.PrintWriter;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.apache.tuweni.bytes.Bytes32;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Helper class for counting and pruning trie logs */
+public class TrieLogHelper {
+  private static final String TRIE_LOG_FILE = "trieLogsToRetain";
+  private static final long BATCH_SIZE = 20_000;
+  private static final int ROCKSDB_MAX_INSERTS_PER_TRANSACTION = 1000;
+  private static final Logger LOG = LoggerFactory.getLogger(TrieLogHelper.class);
+
+  static void prune(
+      final DataStorageConfiguration config,
+      final BonsaiWorldStateKeyValueStorage rootWorldStateStorage,
+      final MutableBlockchain blockchain,
+      final Path dataDirectoryPath) {
+    final String batchFileNameBase = dataDirectoryPath.resolve(DATABASE_PATH) + "/" + TRIE_LOG_FILE;
+    TrieLogHelper.validatePruneConfiguration(config);
+
+    final long layersToRetain = config.getUnstable().getBonsaiTrieLogRetentionThreshold();
+
+    final long chainHeight = blockchain.getChainHeadBlockNumber();
+
+    final long lastBlockNumberToRetainTrieLogsFor = chainHeight - layersToRetain + 1;
+
+    if (!validPruneRequirements(blockchain, chainHeight, lastBlockNumberToRetainTrieLogsFor)) {
+      return;
+    }
+
+    final long numberOfBatches = calculateNumberofBatches(layersToRetain);
```

Comment:
Why does this need to be done in batches?

Reply:
I'm using limited batches to provide a safety allocation of resources. Mostly a safety limit to prevent besu to get killed or exhaust resources if a user decide to retains a high number of trie logs.